File: gtci/08-monotonic-stacks/python3/2-remove-all-adjacent-duplicates-in-string.py
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack = []

        for ch in s:
            if len(stack) > 0 and ch == stack[-1]:
                stack.pop()
            else:
                stack.append(ch)

        return "".join(stack)


if __name__ == "__main__":
    sol = Solution()

    assert sol.removeDuplicates("abccba") == ""
    assert sol.removeDuplicates("foobar") == "fbar"
    assert sol.removeDuplicates("fooobar") == "fobar"
    assert sol.removeDuplicates("abcd") == "abcd"

    print("All test cases passed.")

File: gtci/08-monotonic-stacks/cpp/2-remove-all-adjacent-duplicates-in-string.cpp
#include <cassert>
#include <iostream>
#include <stack>

class Solution {
   public:
    std::string removeDuplicates(std::string s) {
        std::stack<char> charStack;

        for (char ch : s) {
            if (charStack.size() > 0 && ch == charStack.top())
                charStack.pop();
            else
                charStack.push(ch);
        }

        std::string result = "";
        while (charStack.size() > 0) {
            result = charStack.top() + result;
            charStack.pop();
        }

        return result;
    }
};

int main() {
    Solution sol = Solution();

    assert(sol.removeDuplicates("abccba") == "");
    assert(sol.removeDuplicates("foobar") == "fbar");
    assert(sol.removeDuplicates("fooobar") == "fobar");
    assert(sol.removeDuplicates("abcd") == "abcd");

    std::cout << "All test cases passed." << std::endl;

    return 0;
}
